### PR TITLE
TYP: Fix mypy ignores in parsers

### PR DESCRIPTION
**Pull request by phofl**
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them

Tried fixing the Mypy Errors in parsers

### CODE REVIEW

**Comment by jreback** [comment]
can you merge master

**Comment by phofl** [comment]
Done

@simonjayhawkins Is it ok to use the | syntax?

**Comment by MarcoGorelli** [comment]
quick comment on the `|` syntax - this is in the pyupgrade roadmap (https://github.com/asottile/pyupgrade/issues/372)

**Review by jreback — APPROVED** [review]
wow this looks good. cc @simonjayhawkins

File: pandas/io/parsers/base_parser.py [hunk 1]
```diff
@@ -335,9 +348,7 @@ def _maybe_dedup_names(self, names):
         # would be nice!
         if self.mangle_dupe_cols:
             names = list(names)  # so we can index
-            # pandas\io\parsers.py:1559: error: Need type annotation for
-            # 'counts'  [var-annotated]
-            counts = defaultdict(int)  # type: ignore[var-annotated]
+            counts: DefaultDict[int | str, int] = defaultdict(int)
```
Comment (simonjayhawkins) [inline pandas/io/parsers/base_parser.py:351]: if an mi, the dict key is a tuple?

Comment (simonjayhawkins) [inline pandas/io/parsers/base_parser.py:351]: if an mi, the dict key is a tuple?

Comment (phofl) [reply pandas/io/parsers/base_parser.py]: You are right, thx. I looked them through again and fixed 2 more I think

**Comment by phofl** [comment]
ping @simonjayhawkins

File: pandas/io/parsers/base_parser.py [hunk 2]
```diff
@@ -549,7 +558,7 @@ def _convert_to_ndarrays(
         return result
 
     def _set_noconvert_dtype_columns(
-        self, col_indices: List[int], names: List[Union[int, str]]
+        self, col_indices: List[int], names: List[int | str | tuple]
```
Comment (MarcoGorelli) [inline pandas/io/parsers/base_parser.py:561]: see #39538 for some discussion on this syntax change - perhaps it's best to defer it?

Comment (phofl) [reply pandas/io/parsers/base_parser.py]: Rolled back

**Comment by jreback** [comment]
thanks @phofl

**Pull request merged**